src/sampling.jl

```diff
@@ -206,6 +206,32 @@ end
 self_avoid_sample!(a::AbstractArray, x::AbstractArray) =
     self_avoid_sample!(Base.GLOBAL_RNG, a, x)
 
+"""
+    shrinking_array_sample!([rng], a::AbstractArray, x::AbstractArray)
+
+Shrinking-array sampling: samples from an array by replacing each selected
+element with the last element in the array and decrementing the array length
+counter by 1 each time - hence shrinking-array.
+"""
+function shrinking_array_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
+    n = length(a)
+    k = length(x)
+    k <= n || throw(DimensionMismatch("length(x) should not exceed length(a)"))
+
+    # need to use an index instead as in the case where a = 1:j then a[jj] = val
```

Comment:
What is `jj`?

Reply:
`jj` is just an arbitrary integer. Should make more clear. The point is `a[jj]` for arbitrary integer `jj` will fail if `a` is a `UnitRange` . Basically, when `a` doesn't have a `setindex!` but is still `typeof(a) <: AbstractArray` then the original code will fail, hence this fix.